install.sh

```diff
@@ -112,108 +229,120 @@ _install_mac() {
-      source "$HOME/.cargo/env"
+    [Yy]*|"")
+      curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y || _exit_failure "install-rust"
+      source "$HOME/.cargo/env" # TODO: this only affects this script's env. Print hint to do this manually after installation for the user's shell
+      break
+      ;;
+    [Nn]*)
+      _exit_neutral
       ;;
-    [Nn]*) exit ;;
     *) echo "Please answer yes or no." ;;
     esac
   done
 }
 
 _install_with_cargo() {
   echo "Installing with cargo..."
-  cargo install --locked cargo-shuttle
+  INSTALL_METHOD="cargo"
+  cargo install --locked cargo-shuttle || _exit_failure "cargo-install"
 }
 
 _install_default() {
   echo "Installing with package manager is not supported"
 
   if command -v cargo-binstall &>/dev/null; then
     echo "Installing with cargo-binstall"
-    cargo binstall -y --locked cargo-shuttle
+    INSTALL_METHOD="cargo-binstall"
+    cargo-binstall -y --force --locked cargo-shuttle || _exit_failure "cargo-binstall"
     return 0
   fi
 
-  if ! command -v cargo &>/dev/null; then
-    if ! command -v rustup &>/dev/null; then
-      _install_rust_and_cargo
-    else
-      echo "rustup was found, but cargo wasn't. Something is up with your install"
-      exit 1
-    fi
-  fi
-
   while true; do
-    read -r -p "Do you wish to attempt to install the pre-built binary? [Y/N] " yn </dev/tty
+    read -r -p "Install the pre-built binary? [Y/n] " yn </dev/tty
     case $yn in
-    [Yy]*)
-      echo "Installing pre-built binary..."
+    [Yy]*|"")
       _install_binary
       break
       ;;
     [Nn]*)
-      read -r -p "Do you wish to attempt an install with cargo? [Y/N] " yn </dev/tty
-      case $yn in
-      [Yy]*)
-        _install_with_cargo
-        break
-        ;;
-      [Nn]*) exit ;;
-      *) echo "Please answer yes or no." ;;
-      esac
+      while true; do
+        read -r -p "Install from source with cargo? [Y/n] " yn </dev/tty
+        case $yn in
+        [Yy]*|"")
+          _install_with_cargo
+          break
+          ;;
+        [Nn]*)
+          _exit_neutral
+          ;;
+        *) echo "Please answer yes or no." ;;
+        esac
+      done
       ;;
     *) echo "Please answer yes or no." ;;
     esac
   done
 }
 
-if command -v cargo-shuttle &>/dev/null; then
-  if [[ "$(cargo-shuttle -V)" = *"${LATEST_VERSION#v}" ]]; then
-    echo "cargo-shuttle is already at the latest version!"
-    exit
+# Check Rust installation since it is a required dependency
+if ! command -v cargo &>/dev/null; then
+  if ! command -v rustup &>/dev/null; then
+    _install_rust_and_cargo
+    echo
   else
-    echo "Updating cargo-shuttle to $LATEST_VERSION"
+    echo "rustup was found, but cargo wasn't. Something is up with your Rust installation."
+    _exit_failure "rustup-found-cargo-not-found"
   fi
 fi
 
 case "$OSTYPE" in
-linux*) _install_linux ;;
-darwin*) _install_mac ;;
-*) _install_default ;;
+linux*)
+  PLATFORM="linux"
```

Comment:
We need to perform platform detection at the top of the script, and do it as early as possible, so that the OS information is recorded in any failures that may happen early in the script.